File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int cnt=0;
        int l=0;
        unordered_map<char, int> ch_cnt = {{'a', 0}, {'b', 0}, {'c', 0}};

        for(int r=0; r<s.size(); r++){
            ch_cnt[s[r]]++;

            while(ch_cnt['a']>0 && ch_cnt['b']>0 && ch_cnt['c']>0){
                cnt += s.length() - r;
                ch_cnt[s[l]]--;
                l++;
            }
        }

        return cnt;
    }
};

File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        last_seen = [-1, -1, -1]
        cnt = 0
        
        for i, char in enumerate(s):
            last_seen[ord(char) - ord('a')] = i
            if last_seen[0] != -1 and last_seen[1] != -1 and last_seen[2] != -1:
                cnt += 1 + min(last_seen)
        
        return cnt